eventsources/sources/redisstream/start.go

```diff
@@ -0,0 +1,178 @@
+limitations under the License.
+*/
+
+package redisstream
+
+import (
+	"context"
+	"encoding/json"
+	"strings"
+	"time"
+
+	"github.com/go-redis/redis"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+
+	"github.com/argoproj/argo-events/common"
+	"github.com/argoproj/argo-events/common/logging"
+	"github.com/argoproj/argo-events/eventsources/sources"
+	metrics "github.com/argoproj/argo-events/metrics"
+	apicommon "github.com/argoproj/argo-events/pkg/apis/common"
+	"github.com/argoproj/argo-events/pkg/apis/events"
+	"github.com/argoproj/argo-events/pkg/apis/eventsource/v1alpha1"
+)
+
+// EventListener implements Eventing for the Redis event source
+type EventListener struct {
+	EventSourceName string
+	EventName       string
+	EventSource     v1alpha1.RedisStreamEventSource
+	Metrics         *metrics.Metrics
+}
+
+// GetEventSourceName returns name of event source
+func (el *EventListener) GetEventSourceName() string {
+	return el.EventSourceName
+}
+
+// GetEventName returns name of event
+func (el *EventListener) GetEventName() string {
+	return el.EventName
+}
+
+// GetEventSourceType return type of event server
+func (el *EventListener) GetEventSourceType() apicommon.EventSourceType {
+	return apicommon.RedisStreamEvent
+}
+
+// StartListening listens for new data on specified redis streams
+func (el *EventListener) StartListening(ctx context.Context, dispatch func([]byte) error) error {
+	log := logging.FromContext(ctx).
+		With(logging.LabelEventSourceType, el.GetEventSourceType(), logging.LabelEventName, el.GetEventName())
+	log.Info("started processing the Redis stream event source...")
+	defer sources.Recover(el.GetEventName())
+
+	redisEventSource := &el.EventSource
+
+	opt := &redis.Options{
+		Addr: redisEventSource.HostAddress,
+		DB:   int(redisEventSource.DB),
+	}
+
+	log.Info("retrieving password if it has been configured...")
+	if redisEventSource.Password != nil {
+		password, err := common.GetSecretFromVolume(redisEventSource.Password)
+		if err != nil {
+			return errors.Wrapf(err, "failed to find the secret password %s", redisEventSource.Password.Name)
+		}
+		opt.Password = password
+	}
+
+	if redisEventSource.TLS != nil {
+		tlsConfig, err := common.GetTLSConfig(redisEventSource.TLS)
+		if err != nil {
+			return errors.Wrap(err, "failed to get the tls configuration")
+		}
+		opt.TLSConfig = tlsConfig
+	}
+
+	log.Infof("setting up a redis client for %s...", redisEventSource.HostAddress)
+	client := redis.NewClient(opt)
+
+	if status := client.Ping(); status.Err() != nil {
+		return errors.Wrapf(status.Err(), "failed to connect to host %s and db %d for event source %s", redisEventSource.HostAddress, redisEventSource.DB, el.GetEventName())
+	}
+	log.Infof("connected to redis server %s", redisEventSource.HostAddress)
+
+	// Create a common consumer group on all streams.
+	// Only proceeds if all the streams are already present
+	consumersGroup := "argo-events-cg"
```

Comment:
Hmmm. I think I'm ok with creating the consumer group in the code for now, let's not expose `createConsumerGroup`.

I think you still need to consider following:

1. Auto generate consumer group if it's not provided in the spec;
2. When you create the consumerGroup, set the offset to be the latest instead of beginning

Also, since Redis can not rebalance the messages automatically when a client dies, I think it's ok to only use one consumer to process the messages. Just one question, will it be a problem if there are multiple clients trying to connect with same consumer? This happens during a pod restart.

You will also need to add the event source type to https://github.com/argoproj/argo-events/blob/master/pkg/apis/common/common.go#L58.

Reply:
Thanks for the feedback. I will continue working on this from this Friday.

Reply:
Auto generating consumergroup might be a bit tricky. Since its only purpose here is to resume from last read upon pod restarts, we will need a name that stays constant across pod restarts. The name of the redis stream event source deployment will be a good choice but I'm not sure what is the proper way to get it from the pod (prefix of the pod's hostname might work). Should I just make this option mandatory ?

If 2 pods use same consumergroup and same consumer name to read from a redis stream, only one of them will see the new entry. Tested this by
```sh
XGROUP CREATE mystream mygroup $

# Ran this in two clients. Both blocks with same consumergroup and consumer until new message arrives.
# Only one client sees the new message.
XREADGROUP BLOCK 0 GROUP mygroup Alice STREAMS mystream >
```

Event source is already added to RecreateStrategyEventSources https://github.com/BulkBeing/argo-events/blob/e9abeb3896e235e9e53b31c8ac0ef115e333e4c3/pkg/apis/common/common.go#L72